### ISSUE TITLE
@interact signature

### ISSUE THREAD

**Issue by Carreau** [ipython/ipywidgets#430]
Discussed with @fperez and @SylvainCorlay to try to fix (some) pain with `@interact`. 
Basically we **need** `@interact` to have an `(*args, **kwarg)` signature ; but we want to be able to pass some options to interact. Typically, that's why we have `interact_manual()`, and now we want to be able to select wether or not to make the output into a widget [example Lorenz attractor have a huge output].

Basically we could have `@interact` being a callable object, (which `__call__` being the current version of `@interact` and an `option` methods to set various options (and return self for chaining).

The usage would thus look like:

```
@interact.option(manual=True, output=False) 
def function(...):
      pass
```

or 

```
@interact.manual(True).output(False) 
def function(...):
      pass
```

or a like.

**Comment by ellisonbg** [comment]
How would this API work if you are using it with a function that was
already defined by someone else?

On Thu, Mar 17, 2016 at 12:49 PM, Matthias Bussonnier <
notifications@github.com> wrote:

> Discussed with @fperez <https://github.com/fperez> and @SylvainCorlay
> <https://github.com/SylvainCorlay> to try to fix (some) pain with
> @interact.
> Basically we *need* @interact to have an (*args, **kwarg) signature ; but
> we want to be able to pass some options to interact. Typically, that's why
> we have interact_manual(), and now we want to be able to select wether or
> not to make the output into a widget [example Lorenz attractor have a huge
> output].
>
> Basically we could have @interact being a callable object, (which __call__
> being the current version of @interact and an option methods to set
> various options (and return self for chaining).
>
> The usage would thus look like:
>
> @interact.option(manual=True, output=False)
> def function(...):
>       pass
>
> or
>
> @interact.manual(True).output(False)
> def function(...):
>       pass
>
> or a like.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipywidgets/issues/430>
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by fperez** [comment]
It would be fine: `interact(f, x=True, y=(1,10)).options(manual=True, output=False)`.  The idea is to construct a callable that's a very thin wrapper around the passed function, and which uses the 'return self' pattern in its methods to allow method-chaining.

It would be a change of the current API, in that we wouldn't return the original function/method but instead a wrapper around it.  That's the only problem with this approach.

An alternative is to use instead: `interact.options(...)(f, x=True)`.  That would look a bit wonky, but it would allow us to preserve the current return type API.  I'm leaning towards this second idea, b/c `f` could be a complex callable object (not just a function), so the "wrap it into a callable" can get a little messy.

**Comment by ellisonbg** [comment]
Like this idea very much!

On Thu, Mar 17, 2016 at 2:10 PM, Fernando Perez <notifications@github.com>
wrote:

> It would be fine: interact(f, x=True, y=(1,10)).options(manual=True,
> output=False). The idea is to construct a callable that's a very thin
> wrapper around the passed function, and which uses the 'return self'
> pattern in its methods to allow method-chaining.
>
> It would be a change of the current API, in that we wouldn't return the
> original function/method but instead a wrapper around it. That's the only
> problem with this approach.
>
> An alternative is to use instead: interact.options(...)(f, x=True). That
> would look a bit wonky, but it would allow us to preserve the current
> return type API. I'm leaning towards this second idea, b/c f could be a
> complex callable object (not just a function), so the "wrap it into a
> callable" can get a little messy.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipywidgets/issues/430#issuecomment-198083779>
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by SylvainCorlay** [comment]
:+1: This is a great idea.

What do you think the API should look like when not using the decorator but the direct API?

**Comment by fperez** [comment]
See my comment above...

**Comment by SylvainCorlay** [comment]
> See my comment above...

I was thinking of the `interactive` object that is public API.

**Comment by fperez** [comment]
Ah, actually I think for `interactive` it's *easier*: since we already return an object, we can make it have an `.options()` method to do the same thing. And it's easy to make the API symmetric, so that the top-level calls both support `interact{ive}.options(...)(f, ...)`, but have the *return* value of `interactive` also maintain the same `options` method afterwards.

Does that sound right?

**Comment by ellisonbg** [comment]
I think that sounds good for both APIs

On Thu, Mar 17, 2016 at 2:40 PM, Fernando Perez <notifications@github.com>
wrote:

> Ah, actually I think for interactive it's *easier*: since we already
> return an object, we can make it have an .options() method to do the same
> thing. And it's easy to make the API symmetric, so that the top-level calls
> both support interact{ive}.options(...)(f, ...), but have the *return*
> value of interactive also maintain the same options method afterwards.
>
> Does that sound right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipywidgets/issues/430#issuecomment-198094201>
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by SylvainCorlay** [comment]
Yes it makes sense.

**Comment by SylvainCorlay** [comment]
cc @jdemeyer

**Comment by jdemeyer** [comment]
I like the original proposal
```
@interact.option(manual=True, output=False)
```
best.

The problem with `interact(f).option(...)` is that you're first creating the interact one way and then want to change it. That's harder to do (but not impossible).

**Comment by jdemeyer** [comment]
Actually, let me work on this...

**Comment by jdemeyer** [comment]
I just found out that Python's syntax won't allow complex decorators like
```
@interact.option(manual=True)(x=20)
```

I haven't found the precise syntax definition of decorators, but it seems limited to one function call. Note that this is a purely **syntactical** issue.

This gives additional rationale for the `interact.option` syntax, as opposed to `interact.manual(True)`.

**Comment by jdemeyer** [comment]
Done: #771

**Comment by jdemeyer** [comment]
Found the decorator syntax [here](https://docs.python.org/2/reference/compound_stmts.html#function-definitions)
```
decorator      ::=  "@" dotted_name ["(" [argument_list [","]] ")"] NEWLINE
```
So it indeed allows only a particular simple syntax.

**Comment by Carreau** [comment]
> Found the decorator syntax here
> 
> decorator      ::=  "@" dotted_name ["(" [argument_list [","]] ")"] NEWLINE
> So it indeed allows only a particular simple syntax.

:-( that make me sad. 

I'm wondering if we should open an issue to note that and see if we can make a pep for Python 3.7.

**Comment by jasongrout** [comment]
From the [PEP](https://www.python.org/dev/peps/pep-0318/#current-syntax):

> The decorator statement is limited in what it can accept -- arbitrary expressions will not work. Guido preferred this because of a [gut feeling](https://mail.python.org/pipermail/python-dev/2004-August/046711.html)

**Comment by jdemeyer** [comment]
Funny. I think only the BDFL could get away with doing something just because of a "gut feeling".

**Comment by Carreau** [comment]
> Funny. I think only the BDFL could get away with doing something just because of a "gut feeling".

I believe that.


> From the PEP:
> 
> The decorator statement is limited in what it can accept -- arbitrary expressions will not work. Guido preferred this because of a gut feeling

Hum I wonder if we could come up with a clear use case and make the CoreDev change their mind.

**Comment by jdemeyer** [comment]
I think this can be closed now (see #771).

**Comment by jasongrout** [comment]
Thanks @jdmeyer!
